##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [2]:
import pandas as pd
import numpy as np

In [22]:
data = pd.read_csv('heart_pandas.csv')
print(data.shape)
print(data.head())
print('---'*30)

(303, 14)
   age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                   

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

#### Сколько мужчин в датасете? Сколько женщен? (sex)

In [26]:
data.sex.value_counts()

sex
male      207
female     96
Name: count, dtype: int64

#### Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) 

In [29]:
print(f"{100 * sum(data['sex'] == 'male') / len(data):.1f}%")

68.3%


Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца?

In [30]:
data.groupby('sex')['target'].sum()

sex
female    72
male      93
Name: target, dtype: int64

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [44]:
men_no_disease = sum((data['sex'] == 'male') & (data['target'] == 0))
total_patients = len(data)

print(f"{men_no_disease/total_patients:.1%}")

37.6%


**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [43]:
data[data['target'] == 1]['age'].min()

np.int64(29)

**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [41]:
data[data['target'] == 0]['age'].max()

np.int64(77)

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [45]:
data[(data['sex'] == 'female') & (data['target'] == 1)]['age'].min()

np.int64(34)

**Какой средний возраст женщин?**

In [51]:
data[data['sex'] == 'female']['age'].mean().round(0)

np.float64(56.0)

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [53]:
sick_stats = data[data['target'] == 1]['age'].agg(['mean', 'std'])
print(f"Пациенты с болезнью сердца:")
print(f"Средний возраст: {sick_stats['mean']:.1f} лет")
print(f"Стандартное отклонение: {sick_stats['std']:.1f} лет")

Пациенты с болезнью сердца:
Средний возраст: 52.5 лет
Стандартное отклонение: 9.6 лет


#### Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) 

In [55]:
# Средний chol
mean_chol_all = data['cholesterol'].mean()

# Средний chol - без болезни сердца
mean_chol_healthy = data[data['target'] == 0]['cholesterol'].mean()

print(f"Средний уровень холестерина для ВСЕХ пациентов: {mean_chol_all:.1f}")
print(f"Средний уровень холестерина для пациентов БЕЗ болезни сердца: {mean_chol_healthy:.1f}")

Средний уровень холестерина для ВСЕХ пациентов: 246.3
Средний уровень холестерина для пациентов БЕЗ болезни сердца: 251.1


**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [61]:
data.groupby('rest_ecg')[['age', 'resting_blood_pressure', 'cholesterol', 'max_heart_rate_achieved', 'st_depression', 'num_major_vessels']].agg(['max', 'mean'])

age            resting_blood_pressure  \
                             max       mean                    max   
rest_ecg                                                             
ST-T wave abnormality         71  52.914474                    180   
left ventricular hypertrophy  76  61.000000                    180   
normal                        77  55.687075                    200   

                                         cholesterol              \
                                    mean         max        mean   
rest_ecg                                                           
ST-T wave abnormality         129.065789         354  237.269737   
left ventricular hypertrophy  140.500000         327  261.750000   
normal                        134.027211         564  255.142857   

                             max_heart_rate_achieved              \
                                                 max        mean   
rest_ecg                                                           
ST-T wave abnormality                            194  151.960526   
left ventricular hypertrophy                     140  125.750000   
normal                                           202  147.904762   

                             st_depression           num_major_vessels  \
                                       max      mean               max   
rest_ecg                                                                 
ST-T wave abnormality                  5.6  0.879605                 4   
left ventricular hypertrophy           4.4  2.725000                 3   
normal                                 6.2  1.159184                 4   

                                        
                                  mean  
rest_ecg                                
ST-T wave abnormality         0.638158  
left ventricular hypertrophy  1.000000  
normal                        0.816327

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [62]:
men_sick_mean = data[(data['sex'] == 'male') & (data['target'] == 1)]['st_depression'].mean()

women_healthy_mean = data[(data['sex'] == 'female') & (data['target'] == 0)]['st_depression'].mean()

print(f"Средний уровень депрессии ST у мужчин С болезнью сердца: {men_sick_mean:.3f}")
print(f"Средний уровень депрессии ST у женщин БЕЗ болезни сердца: {women_healthy_mean:.3f}")

Средний уровень депрессии ST у мужчин С болезнью сердца: 0.605
Средний уровень депрессии ST у женщин БЕЗ болезни сердца: 1.842


**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [65]:
columns_to_analyze = ['chest_pain_type', 'rest_ecg', 'thalassemia']

for column in columns_to_analyze:
    print(f"\n\nСтатистика холестерина по группам '{column}':")
    
    result = data.groupby(column)['cholesterol'].agg(['min', 'max'])
    print(result)



Статистика холестерина по группам 'chest_pain_type':
                  min  max
chest_pain_type           
atypical angina   126  564
non-anginal pain  182  298
typical angina    131  409


Статистика холестерина по группам 'rest_ecg':
                              min  max
rest_ecg                              
ST-T wave abnormality         126  354
left ventricular hypertrophy  197  327
normal                        149  564


Статистика холестерина по группам 'thalassemia':
                   min  max
thalassemia                
fixed defect       141  417
normal             169  318
reversable defect  126  564


**Сколько значений может принимать каждый из категориальных признаков?**

In [67]:
cat_columns = ['sex', 'chest_pain_type', 'fasting_blood_sugar', 'rest_ecg', 
                       'exercise_induced_angina', 'st_slope', 'thalassemia', 'target']

for col in cat_columns:
    unique_values = data[col].nunique()
    values_list = data[col].unique()
    print(f"{col}:")
    print(f"\tКоличество уникальных значений: {unique_values}")
    print(f"\tЗначения: {sorted(values_list) if isinstance(values_list[0], (int, float)) else values_list}")
    print("\n")

sex:
	Количество уникальных значений: 2
	Значения: ['male' 'female']


chest_pain_type:
	Количество уникальных значений: 3
	Значения: ['non-anginal pain' 'atypical angina' 'typical angina']


fasting_blood_sugar:
	Количество уникальных значений: 2
	Значения: ['greater than 120mg/ml' 'lower than 120mg/ml']


rest_ecg:
	Количество уникальных значений: 3
	Значения: ['normal' 'ST-T wave abnormality' 'left ventricular hypertrophy']


exercise_induced_angina:
	Количество уникальных значений: 2
	Значения: ['no' 'yes']


st_slope:
	Количество уникальных значений: 2
	Значения: ['upsloping' 'flat']


thalassemia:
	Количество уникальных значений: 3
	Значения: ['normal' 'fixed defect' 'reversable defect']


target:
	Количество уникальных значений: 2
	Значения: [1 0]




**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [73]:
cat_columns = ['sex', 'chest_pain_type', 'fasting_blood_sugar', 'rest_ecg', 
                       'exercise_induced_angina', 'st_slope', 'thalassemia', 'target']

max_imbalance = 0
most_imbalanced_col = None
imbalance_info = {}

print("Дисбаланс классов по категориальным признакам:")
print("--------")

for col in cat_columns:
    value_counts = data[col].value_counts(normalize=True)
    
    imbalance_ratio = value_counts.max() / value_counts.min()
    
    imbalance_info[col] = {
        'ratio': imbalance_ratio,
        'max_class': value_counts.idxmax(),
        'max_percent': value_counts.max() * 100,
        'min_class': value_counts.idxmin(),
        'min_percent': value_counts.min() * 100,
        'distribution': value_counts
    }
    
    print(f"\n{col}:")
    print(f"\tСоотношение: {imbalance_ratio:.2f}x") # во сколько раз самый частый класс больше самого редкого класса
    print(f"\tfreq: '{value_counts.idxmax()}' - {value_counts.max():.1%}")
    print(f"\trare: '{value_counts.idxmin()}' - {value_counts.min():.1%}")
    
    if imbalance_ratio > max_imbalance:
        max_imbalance = imbalance_ratio
        most_imbalanced_col = col

print("\n\n")
print(f"Топ1 дисбаланс у класса: {most_imbalanced_col}")
print(f"Соотношение классов: {max_imbalance:.2f}x")
print(f"Распределение:")
for value, percent in imbalance_info[most_imbalanced_col]['distribution'].items():
    print(f"  '{value}': {percent:.1%}")

Дисбаланс классов по категориальным признакам:
--------

sex:
	Соотношение: 2.16x
	freq: 'male' - 68.3%
	rare: 'female' - 31.7%

chest_pain_type:
	Соотношение: 8.39x
	freq: 'typical angina' - 63.7%
	rare: 'non-anginal pain' - 7.6%

fasting_blood_sugar:
	Соотношение: 5.73x
	freq: 'lower than 120mg/ml' - 85.1%
	rare: 'greater than 120mg/ml' - 14.9%

rest_ecg:
	Соотношение: 38.00x
	freq: 'ST-T wave abnormality' - 50.2%
	rare: 'left ventricular hypertrophy' - 1.3%

exercise_induced_angina:
	Соотношение: 2.06x
	freq: 'no' - 67.3%
	rare: 'yes' - 32.7%

st_slope:
	Соотношение: 1.13x
	freq: 'upsloping' - 53.1%
	rare: 'flat' - 46.9%

thalassemia:
	Соотношение: 8.30x
	freq: 'fixed defect' - 54.8%
	rare: 'normal' - 6.6%

target:
	Соотношение: 1.20x
	freq: '1' - 54.5%
	rare: '0' - 45.5%



Топ1 дисбаланс у класса: rest_ecg
Соотношение классов: 38.00x
Распределение:
  'ST-T wave abnormality': 50.2%
  'normal': 48.5%
  'left ventricular hypertrophy': 1.3%


---